Error on executing "DELETE FROM table1 WHERE ID = ?; DELETE FROM table2 WHERE ID = ?;"

Executing DELETE FROM SQL that contains two SQL sentences：
`DELETE FROM user_info WHERE ID = ?;
    DELETE FROM role_info WHERE ID = ?;`
This is the relevant stack trace:
`Caused by: org.h2.jdbc.JdbcSQLException: Invalid value "2" for parameter "parameterIndex" [90008-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.getInvalidValueException(DbException.java:240)
	at org.h2.jdbc.JdbcPreparedStatement.setParameter(JdbcPreparedStatement.java:1560)
	at org.h2.jdbc.JdbcPreparedStatement.setString(JdbcPreparedStatement.java:426)
	`

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #1590: Error on executing "DELETE FROM table1 WHERE ID = ?; DELETE FROM table2 WHERE ID = ?;"
+</li>
 <li>Issue #1727: Support ISODOW as identifier for the extract function additional to ISO_DAY_OF_WEEK
 </li>
 <li>PR #1580, #1726: Disable remote database creation by default

--- a/h2/src/main/org/h2/command/CommandContainer.java
+++ b/h2/src/main/org/h2/command/CommandContainer.java
@@ -185,4 +185,8 @@ public class CommandContainer extends Command {
         return prepared.getType();
     }
 
+    void clearCTE() {
+        clearCTE(session, prepared);
+    }
+
 }

--- a/h2/src/main/org/h2/command/CommandList.java
+++ b/h2/src/main/org/h2/command/CommandList.java
@@ -17,13 +17,19 @@ import org.h2.result.ResultInterface;
  */
 class CommandList extends Command {
 
-    private final ArrayList<CommandContainer> commands;
+    private CommandContainer command;
+    private final ArrayList<Prepared> commands;
     private final ArrayList<Parameter> parameters;
+    private String remaining;
+    private Command remainingCommand;
 
-    CommandList(Session session, String sql, ArrayList<CommandContainer> commands, ArrayList<Parameter> parameters) {
+    CommandList(Session session, String sql, CommandContainer command, ArrayList<Prepared> commands,
+            ArrayList<Parameter> parameters, String remaining) {
         super(session, sql);
+        this.command = command;
         this.commands = commands;
         this.parameters = parameters;
+        this.remaining = remaining;
     }
 
     @Override
@@ -32,38 +38,58 @@ class CommandList extends Command {
     }
 
     private void executeRemaining() {
-        for (int i = 1, l = commands.size(); i < l; i++) {
-            CommandContainer command = commands.get(i);
-            if (command.isQuery()) {
-                command.query(0);
+        for (Prepared prepared : commands) {
+            prepared.prepare();
+            if (prepared.isQuery()) {
+                prepared.query(0);
             } else {
-                command.update();
+                prepared.update();
+            }
+        }
+        if (remaining != null) {
+            remainingCommand = session.prepareLocal(remaining);
+            remaining = null;
+            if (remainingCommand.isQuery()) {
+                remainingCommand.query(0);
+            } else {
+                remainingCommand.update();
             }
         }
     }
 
     @Override
     public int update() {
-        int updateCount = commands.get(0).executeUpdate(false).getUpdateCount();
+        int updateCount = command.executeUpdate(false).getUpdateCount();
         executeRemaining();
         return updateCount;
     }
 
     @Override
     public void prepareJoinBatch() {
-        commands.get(0).prepareJoinBatch();
+        command.prepareJoinBatch();
     }
 
     @Override
     public ResultInterface query(int maxrows) {
-        ResultInterface result = commands.get(0).query(maxrows);
+        ResultInterface result = command.query(maxrows);
         executeRemaining();
         return result;
     }
 
     @Override
+    public void stop() {
+        command.stop();
+        for (Prepared prepared : commands) {
+            CommandContainer.clearCTE(session, prepared);
+        }
+        if (remainingCommand != null) {
+            remainingCommand.stop();
+        }
+    }
+
+    @Override
     public boolean isQuery() {
-        return commands.get(0).isQuery();
+        return command.isQuery();
     }
 
     @Override
@@ -78,12 +104,12 @@ class CommandList extends Command {
 
     @Override
     public ResultInterface queryMeta() {
-        return commands.get(0).queryMeta();
+        return command.queryMeta();
     }
 
     @Override
     public int getCommandType() {
-        return commands.get(0).getCommandType();
+        return command.getCommandType();
     }
 
 }

--- a/h2/src/main/org/h2/command/CommandList.java
+++ b/h2/src/main/org/h2/command/CommandList.java
@@ -17,10 +17,10 @@ import org.h2.result.ResultInterface;
  */
 class CommandList extends Command {
 
-    private final ArrayList<Command> commands;
+    private final ArrayList<CommandContainer> commands;
     private final ArrayList<Parameter> parameters;
 
-    CommandList(Session session, String sql, ArrayList<Command> commands, ArrayList<Parameter> parameters) {
+    CommandList(Session session, String sql, ArrayList<CommandContainer> commands, ArrayList<Parameter> parameters) {
         super(session, sql);
         this.commands = commands;
         this.parameters = parameters;
@@ -33,7 +33,7 @@ class CommandList extends Command {
 
     private void executeRemaining() {
         for (int i = 1, l = commands.size(); i < l; i++) {
-            Command command = commands.get(i);
+            CommandContainer command = commands.get(i);
             if (command.isQuery()) {
                 command.query(0);
             } else {

--- a/h2/src/main/org/h2/command/CommandList.java
+++ b/h2/src/main/org/h2/command/CommandList.java
@@ -8,6 +8,7 @@ package org.h2.command;
 import java.util.ArrayList;
 
 import org.h2.engine.Session;
+import org.h2.expression.Parameter;
 import org.h2.expression.ParameterInterface;
 import org.h2.result.ResultInterface;
 
@@ -16,51 +17,53 @@ import org.h2.result.ResultInterface;
  */
 class CommandList extends Command {
 
-    private final Command command;
-    private final String remaining;
+    private final ArrayList<Command> commands;
+    private final ArrayList<Parameter> parameters;
 
-    CommandList(Session session, String sql, Command c, String remaining) {
+    CommandList(Session session, String sql, ArrayList<Command> commands, ArrayList<Parameter> parameters) {
         super(session, sql);
-        this.command = c;
-        this.remaining = remaining;
+        this.commands = commands;
+        this.parameters = parameters;
     }
 
     @Override
     public ArrayList<? extends ParameterInterface> getParameters() {
-        return command.getParameters();
+        return parameters;
     }
 
     private void executeRemaining() {
-        Command remainingCommand = session.prepareLocal(remaining);
-        if (remainingCommand.isQuery()) {
-            remainingCommand.query(0);
-        } else {
-            remainingCommand.update();
+        for (int i = 1, l = commands.size(); i < l; i++) {
+            Command command = commands.get(i);
+            if (command.isQuery()) {
+                command.query(0);
+            } else {
+                command.update();
+            }
         }
     }
 
     @Override
     public int update() {
-        int updateCount = command.executeUpdate(false).getUpdateCount();
+        int updateCount = commands.get(0).executeUpdate(false).getUpdateCount();
         executeRemaining();
         return updateCount;
     }
 
     @Override
     public void prepareJoinBatch() {
-        command.prepareJoinBatch();
+        commands.get(0).prepareJoinBatch();
     }
 
     @Override
     public ResultInterface query(int maxrows) {
-        ResultInterface result = command.query(maxrows);
+        ResultInterface result = commands.get(0).query(maxrows);
         executeRemaining();
         return result;
     }
 
     @Override
     public boolean isQuery() {
-        return command.isQuery();
+        return commands.get(0).isQuery();
     }
 
     @Override
@@ -75,12 +78,12 @@ class CommandList extends Command {
 
     @Override
     public ResultInterface queryMeta() {
-        return command.queryMeta();
+        return commands.get(0).queryMeta();
     }
 
     @Override
     public int getCommandType() {
-        return command.getCommandType();
+        return commands.get(0).getCommandType();
     }
 
 }

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -3878,6 +3878,7 @@ public class Parser {
             if (p == null) {
                 p = new Parameter(index);
                 indexedParameterList.set(index, p);
+                parameters.add(p);
             }
             read();
         } else {
@@ -3887,8 +3888,8 @@ public class Parser {
                         .get(ErrorCode.CANNOT_MIX_INDEXED_AND_UNINDEXED_PARAMS);
             }
             p = new Parameter(parameters.size());
+            parameters.add(p);
         }
-        parameters.add(p);
         return p;
     }
 

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -673,11 +673,11 @@ public class Parser {
      */
     public Command prepareCommand(String sql) {
         try {
-            Command c = prepareSingleCommand(sql);
+            CommandContainer c = prepareSingleCommand(sql);
             if (currentTokenType == SEMICOLON) {
                 String remaining = originalSQL.substring(parseIndex);
                 if (!StringUtils.isWhitespaceOrEmpty(remaining)) {
-                    c = prepareCommandList(c, sql, remaining);
+                    return prepareCommandList(c, sql, remaining);
                 }
             }
             return c;
@@ -686,19 +686,27 @@ public class Parser {
         }
     }
 
-    private Command prepareCommandList(Command c, String sql, String remaining) {
-        ArrayList<Command> list = Utils.newSmallArrayList();
+    private CommandList prepareCommandList(CommandContainer c, String sql, String remaining) {
+        ArrayList<CommandContainer> list = Utils.newSmallArrayList();
         list.add(c);
         do {
             suppliedParameters = parameters;
             suppliedParameterList = indexedParameterList;
-            list.add(prepareSingleCommand(remaining));
+            try {
+                c = prepareSingleCommand(remaining);
+            } catch (Throwable t) {
+                for (CommandContainer cc : list) {
+                    cc.clearCTE();
+                }
+                throw t;
+            }
+            list.add(c);
         } while (currentTokenType == SEMICOLON
                 && !StringUtils.isWhitespaceOrEmpty(remaining = originalSQL.substring(parseIndex)));
         return new CommandList(session, sql, list, parameters);
     }
 
-    private Command prepareSingleCommand(String sql) {
+    private CommandContainer prepareSingleCommand(String sql) {
         Prepared p = parse(sql);
         if (currentTokenType != SEMICOLON && currentTokenType != END) {
             addExpected(SEMICOLON);

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -634,14 +634,15 @@ public class Parser {
     private Prepared currentPrepared;
     private Select currentSelect;
     private ArrayList<Parameter> parameters;
+    private ArrayList<Parameter> indexedParameterList;
+    private ArrayList<Parameter> suppliedParameters;
+    private ArrayList<Parameter> suppliedParameterList;
     private String schemaName;
     private ArrayList<String> expectedList;
     private boolean rightsChecked;
     private boolean recompileAlways;
     private boolean literalsChecked;
-    private ArrayList<Parameter> indexedParameterList;
     private int orderInFrom;
-    private ArrayList<Parameter> suppliedParameterList;
 
     public Parser(Session session) {
         this.database = session.getDatabase();
@@ -672,28 +673,44 @@ public class Parser {
      */
     public Command prepareCommand(String sql) {
         try {
-            Prepared p = parse(sql);
-            boolean hasMore = isToken(SEMICOLON);
-            if (!hasMore && currentTokenType != END) {
-                throw getSyntaxError();
-            }
-            try {
-                p.prepare();
-            } catch (Throwable t) {
-                CommandContainer.clearCTE(session, p);
-                throw t;
-            }
-            Command c = new CommandContainer(session, sql, p);
-            if (hasMore) {
+            Command c = prepareSingleCommand(sql);
+            if (currentTokenType == SEMICOLON) {
                 String remaining = originalSQL.substring(parseIndex);
                 if (!StringUtils.isWhitespaceOrEmpty(remaining)) {
-                    c = new CommandList(session, sql, c, remaining);
+                    c = prepareCommandList(c, sql, remaining);
                 }
             }
             return c;
         } catch (DbException e) {
             throw e.addSQL(originalSQL);
         }
+    }
+
+    private Command prepareCommandList(Command c, String sql, String remaining) {
+        ArrayList<Command> list = Utils.newSmallArrayList();
+        list.add(c);
+        do {
+            suppliedParameters = parameters;
+            suppliedParameterList = indexedParameterList;
+            list.add(prepareSingleCommand(remaining));
+        } while (currentTokenType == SEMICOLON
+                && !StringUtils.isWhitespaceOrEmpty(remaining = originalSQL.substring(parseIndex)));
+        return new CommandList(session, sql, list, parameters);
+    }
+
+    private Command prepareSingleCommand(String sql) {
+        Prepared p = parse(sql);
+        if (currentTokenType != SEMICOLON && currentTokenType != END) {
+            addExpected(SEMICOLON);
+            throw getSyntaxError();
+        }
+        try {
+            p.prepare();
+        } catch (Throwable t) {
+            CommandContainer.clearCTE(session, p);
+            throw t;
+        }
+        return new CommandContainer(session, sql, p);
     }
 
     /**
@@ -727,12 +744,12 @@ public class Parser {
         } else {
             expectedList = null;
         }
-        parameters = Utils.newSmallArrayList();
+        parameters = suppliedParameters != null ? suppliedParameters : Utils.<Parameter>newSmallArrayList();
+        indexedParameterList = suppliedParameterList;
         currentSelect = null;
         currentPrepared = null;
         createView = null;
         recompileAlways = false;
-        indexedParameterList = suppliedParameterList;
         read();
         return parsePrepared();
     }

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -686,24 +686,44 @@ public class Parser {
         }
     }
 
-    private CommandList prepareCommandList(CommandContainer c, String sql, String remaining) {
-        ArrayList<CommandContainer> list = Utils.newSmallArrayList();
-        list.add(c);
-        do {
-            suppliedParameters = parameters;
-            suppliedParameterList = indexedParameterList;
-            try {
-                c = prepareSingleCommand(remaining);
-            } catch (Throwable t) {
-                for (CommandContainer cc : list) {
-                    cc.clearCTE();
+    private CommandList prepareCommandList(CommandContainer command, String sql, String remaining) {
+        try {
+            ArrayList<Prepared> list = Utils.newSmallArrayList();
+            boolean stop = false;
+            do {
+                if (stop) {
+                    return new CommandList(session, sql, command, list, parameters, remaining);
                 }
-                throw t;
-            }
-            list.add(c);
-        } while (currentTokenType == SEMICOLON
-                && !StringUtils.isWhitespaceOrEmpty(remaining = originalSQL.substring(parseIndex)));
-        return new CommandList(session, sql, list, parameters);
+                suppliedParameters = parameters;
+                suppliedParameterList = indexedParameterList;
+                Prepared p;
+                try {
+                    p = parse(remaining);
+                } catch (DbException ex) {
+                    // This command may depend on results of previous commands.
+                    if (ex.getErrorCode() == ErrorCode.CANNOT_MIX_INDEXED_AND_UNINDEXED_PARAMS) {
+                        throw ex;
+                    }
+                    return new CommandList(session, sql, command, list, parameters, remaining);
+                }
+                if (p instanceof DefineCommand) {
+                    // Next commands may depend on results of this command.
+                    stop = true;
+                }
+                list.add(p);
+                if (currentTokenType == END) {
+                    break;
+                }
+                if (currentTokenType != SEMICOLON) {
+                    addExpected(SEMICOLON);
+                    throw getSyntaxError();
+                }
+            } while (!StringUtils.isWhitespaceOrEmpty(remaining = originalSQL.substring(parseIndex)));
+            return new CommandList(session, sql, command, list, parameters, null);
+        } catch (Throwable t) {
+            command.clearCTE();
+            throw t;
+        }
     }
 
     private CommandContainer prepareSingleCommand(String sql) {

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -673,7 +673,21 @@ public class Parser {
      */
     public Command prepareCommand(String sql) {
         try {
-            CommandContainer c = prepareSingleCommand(sql);
+            Prepared p = parse(sql);
+            if (currentTokenType != SEMICOLON && currentTokenType != END) {
+                addExpected(SEMICOLON);
+                throw getSyntaxError();
+            }
+            try {
+                p.prepare();
+            } catch (Throwable t) {
+                CommandContainer.clearCTE(session, p);
+                throw t;
+            }
+            if (parseIndex < sql.length()) {
+                sql = sql.substring(0, parseIndex);
+            }
+            CommandContainer c = new CommandContainer(session, sql, p);
             if (currentTokenType == SEMICOLON) {
                 String remaining = originalSQL.substring(parseIndex);
                 if (!StringUtils.isWhitespaceOrEmpty(remaining)) {
@@ -724,24 +738,6 @@ public class Parser {
             command.clearCTE();
             throw t;
         }
-    }
-
-    private CommandContainer prepareSingleCommand(String sql) {
-        Prepared p = parse(sql);
-        if (currentTokenType != SEMICOLON && currentTokenType != END) {
-            addExpected(SEMICOLON);
-            throw getSyntaxError();
-        }
-        try {
-            p.prepare();
-        } catch (Throwable t) {
-            CommandContainer.clearCTE(session, p);
-            throw t;
-        }
-        if (parseIndex < sql.length()) {
-            sql = sql.substring(0, parseIndex);
-        }
-        return new CommandContainer(session, sql, p);
     }
 
     /**

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -710,6 +710,9 @@ public class Parser {
             CommandContainer.clearCTE(session, p);
             throw t;
         }
+        if (parseIndex < sql.length()) {
+            sql = sql.substring(0, parseIndex);
+        }
         return new CommandContainer(session, sql, p);
     }
 

--- a/h2/src/test/org/h2/test/jdbc/TestPreparedStatement.java
+++ b/h2/src/test/org/h2/test/jdbc/TestPreparedStatement.java
@@ -199,6 +199,7 @@ public class TestPreparedStatement extends TestDb {
         testColumnMetaDataWithEquals(conn);
         testColumnMetaDataWithIn(conn);
         testValueResultSet(conn);
+        testMultipleStatements(conn);
         conn.close();
         testPreparedStatementWithLiteralsNone();
         testPreparedStatementWithIndexedParameterAndLiteralsNone();
@@ -1756,6 +1757,33 @@ public class TestPreparedStatement extends TestDb {
                 }
             }
         }
+    }
+
+    private void testMultipleStatements(Connection conn) throws SQLException {
+        assertThrows(ErrorCode.CANNOT_MIX_INDEXED_AND_UNINDEXED_PARAMS, conn).prepareStatement("SELECT ?; SELECT ?1");
+        assertThrows(ErrorCode.CANNOT_MIX_INDEXED_AND_UNINDEXED_PARAMS, conn).prepareStatement("SELECT ?1; SELECT ?");
+        Statement stmt = conn.createStatement();
+        stmt.execute("CREATE TABLE TEST (ID IDENTITY, V INT)");
+        PreparedStatement ps = conn.prepareStatement("INSERT INTO TEST(V) VALUES ?; INSERT INTO TEST(V) VALUES ?");
+        ps.setInt(1, 1);
+        ps.setInt(2, 2);
+        ps.executeUpdate();
+        ps = conn.prepareStatement("INSERT INTO TEST(V) VALUES ?2; INSERT INTO TEST(V) VALUES ?1;");
+        ps.setInt(1, 3);
+        ps.setInt(2, 4);
+        ps.executeUpdate();
+        try (ResultSet rs = stmt.executeQuery("SELECT V FROM TEST ORDER BY ID")) {
+            assertTrue(rs.next());
+            assertEquals(1, rs.getInt(1));
+            assertTrue(rs.next());
+            assertEquals(2, rs.getInt(1));
+            assertTrue(rs.next());
+            assertEquals(4, rs.getInt(1));
+            assertTrue(rs.next());
+            assertEquals(3, rs.getInt(1));
+            assertFalse(rs.next());
+        }
+        stmt.execute("DROP TABLE TEST");
     }
 
 }

